lib/inc/internal/sources/dmi_source.hpp

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+namespace whereami { namespace sources {
+
+    /**
+     * Stores information collected via DMI
+     */
+    struct dmi_data
+    {
+        /**
+         * The name of the BIOS vendor
+         * via /sys/class/dmi/id/bios_vendor or dmidecode section 0 vendor
+         */
+        std::string bios_vendor;
+        /**
+         * The board manufacturer
+         * via /sys/class/dmi/id/board_vendor or dmidecode section 2 manufacturer
+         */
+        std::string board_manufacturer;
+        /**
+         * The board product name
+         * via /sys/class/dmi/id/board_name or dmidecode section 2 product name
+         */
+        std::string board_product_name;
+        /**
+         * The system manufacturer (via /sys/class/dmi/id/sys_vendor)
+         * via /sys/class/dmi/id/sys_vendor or dmidecode section 1 manufacturer
+         */
+        std::string manufacturer;
+        /**
+         * The product name
+         * via /sys/class/dmi/id/product_name or dmidecode section 1 product name
+         */
+        std::string product_name;
+    };
+
+    /**
+     * Base DMI data source
+     */
+    class dmi_base
+    {
+    public:
+        dmi_base() {}
+        virtual ~dmi_base() {}
+        /**
+         * Retrieve the BIOS vendor
+         * @return The BIOS vendor's name
+         */
+        std::string bios_vendor() const;
+        /**
+         * Retrieve the board manufacturer
+         * @return The board manufacturer's name
+         */
+        std::string board_manufacturer() const;
+        /**
+         * Retrieve the board product name
+         * @return The board product name
+         */
+        std::string board_product_name() const;
+        /**
+         * Retrieve the system manufacturer
+         * @return The system manufacturer
+         */
+        std::string manufacturer() const;
+        /**
+         * Retrieve the product name
+         * @return The product name
+         */
+        std::string product_name() const;
+    protected:
+        /**
+         * Collected data for this machine based on DMI information
+         */
+        std::unique_ptr<dmi_data> data_;
+    };
+
+    /**
+     * Default DMI data source
+     */
+    class dmi : public dmi_base
+    {
+     public:
+        dmi();
+        virtual ~dmi() {}
+     protected:
+        /**
+         * Attempt to collect virtualization data using DMI
+         */
+        virtual void collect_data();
+        /**
+         * Attempt to collect data from files in /sys/class/dmi/id/
+         * @return Whether any data was collected
+         */
+        virtual bool collect_sys_data();
+        /**
+         * Attempt to collect data from dmidecode executable (requires root)
+         * @return Whether any data was collected
+         */
+        virtual bool collect_dmidecode_data();
```

Comment:
I'm curious - the return types for `collect_data()`, `collect_sys_data()` and `collect_dmidecode_data()` feel a little surprising to me. Given their names I would have anticipated them to return a reference to a data structure containing the data that was collected, or something like that. Does that seem right? If so, then we could even imply that they're going to return an identical or similar data structure by renaming slightly to something like `collect_data_from_sys` and `collect_data_from_dmidecode` or just `data_from_sys` etc... Thoughts?